keps/sig-node/20191211-node-maintenance-lease.md

```diff
@@ -0,0 +1,175 @@
+
+The Lease object should be created automatically, and the ownerRef should
+be the corresponding node so it is removed when the node is deleted.
+
+### User Stories [optional]
+
+#### Story 1
+
+I have a configuration management system that requires the node to be online to perform configuration updates.
+
+I have an automated-snapshot system that stops an instance and takes a snapshot for backup / data retention purposes.
+
+I do not want snapshots to happen during an update.
+
+#### Story 2
+
+I am investigating a problem with the node. I don't want the node to be disrupted
+by automation while I'm performing my investigation or fixes, and I don't want
+to stop the automation components from acting on other nodes.
+
+### Implementation Details/Notes/Constraints
+
+#### Impact to the Kubelet
+
+None. The kubelet does not need to respond to this lease or interact with it in
+any way.  This is for other components to coordination actions they intend to
+take (or prevent) on the node that might be disruptive.
+
+#### Impact to Scheduleability
+
+None. The existence of acquiring of this lease does not imply any taints,
+annotations, labels, cordoning or draining will take place. The Lease object
+itself is not intended as a mechanism to start any automated process.
+
+A user/controller acquires the Lease object, and then that user/controller
+may add/remove taints, cordon or drain, if they choose to.  Other components
+may wish to integrate with the new Lease object to avoid conflict. EG:
+a component that ensures a taint is always present on a node might stop ensuring
+that taint if there is a current valid lease holder, but that implementation
+is out of scope here.
+
+#### Client Behavior
+
+1. Client has required RBAC permissions for new namespace and Lease object
+1. Client checks for the existence of a lease (to be backwards compatible)
+1. If the lease exists, confirm the lease is not currently held by another user
+1. Attempt to acquire the lease and set an appropriate LeaseDurationSeconds
+1. If acquired, perform the necessary operations
+1. Release the lease
+
+### Risks and Mitigations
+
+Clients might not behave appropriately.  We can mitigate this by providing
+a library with an easy to use interface.
+
+## Alternatives
+
+### Annotations on Node or elsewhere
+
+Use a designated annotation on the node object, or possibly
+elsewhere.  Drawbacks to this include excess updates being applied to the node
+object directly, which might have performance implications. Having such an
+annotation natively seems like an anti-pattern and could easily be disrupted
+by existing controllers/clients attempting to manage those annotations.
+
+### Field in Node Spec Or Status
+
+Extending the node API will be much more difficult.  This behavior is also
+somewhat orthogonal to the node itself.
+
+### Use existing system namespace for lease object
+
+We could utilize an existing system namespace for the lease object. The primary
```

Comment:
I'd need to check the updates to the API but the original locking was a patched annotation on the object itself, that may be simpler and eliminate the need to create a separate object.

Reply:
I'm not sure what you mean by patched annotation.

I'm perfectly happy with an annotation on the node.  My preferences are 1) Not a CRD and 2) Present by default (eg, not create an add-on or some other dependency necessary for utilizing this).

Reply:
The original client-go locking used an annotation on a resource (eg a ConfigMap), I think this is what Tim refers to. The implementation underneath would attempt to patch the annotation value with the information about the lease (holder, duration etc) and rely on etcd to ensure conflicts were handled appropriately (ie if you tried to acquire after someone else had updated, you get a 409)